PORT-10112 DORA Metrics guide and implement the necessary demo environment adjustments. (#1551)
port-labs/port-docs: docs/guides/all/setup-dora-metrics.md

    }
  }

```
</details>

Add the mapping config to pagerduty incident [data source](https://app.getport.io/settings/data-sources):
<details>
<summary><b>Incident mapping config (click to expand)</b></summary>

```yaml showLineNumbers
   gitHubRepository:
   combinator: '"and"'
   rules:
     - property: '"$title"'
       operator: '"="'
       value: .title
```
</details>

### Aggregating Incident Data for MTTR

Add the following **aggregation properties** to the **Incident blueprint**

<details>
<summary><b>Aggregation Property JSON (click to expand)</b></summary>

```json showLineNumbers
  "aggregationProperties": {
    "meanTimeToResolve": {
      "title": "Mean Time to Recovery",
      "type": "number",
      "target": "pagerdutyIncident",
      "query": {
        "combinator": "and",
        "rules": [
          {
            "property": "status",
            "operator": "=",
            "value": "resolved"
          }
        ]
      },
      "calculationSpec": {
        "calculationBy": "entities",
        "func": "average",
        "averageOf": "hour",
        "field": "recoveryTime",
        "measureTimeBy": "$createdAt"
      }
    }
  }

```
</details>

> This aggregation setup will calculate the **average MTTR** by considering incidents that have been resolved,

Reviewer: @kodjomiles to clarify, MTTR is "mean time to recovery"
So average is redundant 
```suggestion
This aggregation setup will calculate the **MTTR** by considering incidents that have been resolved,
```

      "query": {
        "combinator": "and",
        "rules": [
          {
            "property": "status",
            "operator": "=",
            "value": "resolved"
          },
          {
            "property": "incidentType",
            "operator": "=",
            "value": "Deployment"
          }
        ]
      },
      "calculationSpec": {
        "func": "count",
        "calculationBy": "entities"
      }
    },
    "total_deployments": {
      "title": "Total Deployments",
      "type": "number",
      "target": "deployment",
      "query": {
        "combinator": "and",
        "rules": [
          {
            "property": "deploymentStatus",
            "operator": "=",
            "value": "Success"
          }
        ]
      },
      "calculationSpec": {
        "func": "count",
        "calculationBy": "entities"
      }
    }
  },
  "relations": {
    "admins": {
      "title": "Admins",
      "target": "githubUser",
      "required": false,
      "many": true
    },
    "githubTeams": {
      "title": "GitHub Teams",
      "target": "githubTeam",
      "required": false,
      "many": true
    }
  }
}
```
</details>

## Visualization

> Let's visualize the DORA metrics using **Port's Dashboard** feature. You can create custom dashboards to track these metrics and monitor your team's performance over time.

Reviewer: ```suggestion
By leveraging Port's Dashboards, you can create custom dashboards to track the metrics and monitor your team's performance over time.
```

### Dashboard Setup
1. Go to **Catalog** in the navigation bar
2. Click **new**
3. Select **New dashboard**
4. Name the dashboard (e.g., DORA Metrics), choose an icon if desired, and click **Create**

This will create a new empty dashboard. Let's get ready-to-add widgets 

### Adding Widgets

<details>
<summary><b>Setup Deployment Frequency Widget</b></summary>

1. Click **+ Widget** and select **Number Chart**.
2. Title: `Deployment Frequency - Monthly`, (add rocket icon, optional).
3. Select `Display single property` and choose **Service** as the **Blueprint**.
4. Select an `Entity` and choose `Monthly Deployment Frequency` as the **Property**.

> 5. Click **Save**.

Reviewer: @kodjomiles maybe we can add for at least one metric a screenshot of how it looks like before "save"?

Author: OKAY

1. Click **+ Widget** and select **Number Chart**.
2. Title: `Lead Time for Changes (Hours)`, (add appropriate icon).
3. Select `Display single property` and choose **Service** as the **Blueprint**.
4. Select an `Entity` and choose `Lead Time for Change` as the **Property**.
5. Click **Save**.

</details>

<details>
<summary><b>Setup Change Failure Rate Widget</b></summary>

1. Click **+ Widget** and select **Number Chart**.
2. Title: `Change Failure Rate`, (add icon, optional).
3. Select `Display single property` and choose **Service** as the **Blueprint**.
4. Select an `Entity` and choose `Change Failure Rate` as the **Property**.
5. Click **Save**.

</details>

:::tip Metric widget groupings
It would be visually cleaner and more informative to group related widgets, such as the **Line Chart** and **Number Chart** widgets, side by side for easier comparison.
You can replicate more examples by checking our dora metrics dashboard on the [demo environment](https://demo.getport.io/dora_metrics).
:::

<img src="/img/guides/doraMetricsDBVisualization.png"/>

By following these steps, you can create a custom dashboard to visualize the DORA metrics.

> Congrats 🎉 You've successfully setup dora metrics in Port 🔥

Reviewer: ```suggestion
Congrats 🎉 You've successfully setup DORA metrics in Port 🔥
```

        "description": "Indicates whether the deployment was successful or failed."
      },
      "leadTime": {
        "title": "Lead Time",
        "type": "number",
        "description": "The time in hours between a pull request being merged and its deployment."
      }
    },
    "aggregationProperties": {
      "deploymentFrequency": {
        "title": "Deployment Frequency",
        "target": "deployment",
        "calculationSpec": {
          "calculationBy": "entities",
          "func": "count"
        },
        "query": {
          "combinator": "and",
          "rules": [
            {
              "property": "createdAt",
              "operator": "within",
              "value": "last_30_days"
            }
          ]
        },
        "description": "Counts the number of deployments over the past 30 days."
      }
    },
    "calculationProperties": {
      "changeFailureRate": {
        "title": "Change Failure Rate",
        "type": "number",
        "calculation": "(.properties.deploymentStatus == 'Failed') / count(.properties.deploymentStatus) * 100",
        "description": "Calculates the percentage of failed deployments."
      }
    }
   },

>   "relations": {

Author: I thought that would be done at the organizational blueprint level, we will add the service relation to the organizational blueprint there, or you want us to add the organizational blueprint here

      "deploymentFrequency": {
        "title": "Deployment Frequency",
        "target": "deployment",
        "calculationSpec": {
          "calculationBy": "entities",
          "func": "count"
        },
        "query": {
          "combinator": "and",
          "rules": [
            {
              "operator": "between",
              "property": "createdAt",
              "value": {
                "preset": "lastMonth"
              }
            }
          ]
        },
        "description": "Counts the number of deployments over the past 30 days."
      }
    },
    
    "calculationProperties": {
      "changeFailureRate": {
        "title": "Change Failure Rate",
        "type": "number",
        "calculation": "(.properties.deploymentStatus == 'Failed') / count(.properties.deploymentStatus) * 100",
        "description": "Calculates the percentage of failed deployments."
      }
    }
   },
  "relations": {
    "service": {
      "title": "Service",
      "target": "service",
      "many": false
    },
    "pullRequest": {
      "title": "Pull Request",
      "target": "pullRequest",
      "many": false
    }
  }
}
```
</details>

---

> ### Metrics Calculation

Author: I am following the order of the notion guide, maybe the heading is misleading
![image](https://github.com/user-attachments/assets/e1997d9f-4f17-4665-8aa3-10579296511e)
maybe you can clarify

> ### Metrics Calculation

Reviewer: @kodjomiles we discussed, metrics can move to another section

- **Using Port’s Calculation Property**:
    - We will introduce a property called **deploymentStatus** to track whether a deployment was **Successful** or **Failed**.
    - The **Change Failure Rate** will be calculated by dividing the number of failed deployments by the total number of deployments in the same period.

  **Steps**:
    - Count all deployments with `deploymentStatus` set to **Failed**.
    - Divide this count by the total number of deployments, and multiply by 100 to get the percentage.

  **Example**:
    - If 2 out of 10 deployments failed, the **Change Failure Rate** would be:

  CFR = (Failed Deployments / Total Deployments) × 100
  CFR = (2 / 10) × 100 = 20%

<details>
<summary><b> Calculation for CFR (click to expand)</b></summary>

```json showLineNumbers
{
  "calculationProperties": {
    "changeFailureRate": {
      "title": "Change Failure Rate",
      "type": "number",
      "calculation": "(.properties.deploymentStatus == 'Failed') / count(.properties.deploymentStatus) * 100"
    }
  }
}
```
</details>

### Tracking Deployment Strategies
The following strategies can help improve your deployment frequency and change failure rate:

> #### PR Merge

Reviewer: Let's add that this is the recommended way

One of the ways to track deployments is by monitoring when pull requests (PRs) are merged into a branch, typically the **main** or **production** branch. The lead time for these merges is calculated as the difference between when the PR was created and when it was merged.

Example:

- When a PR is merged, a **deployment entity** is created in Port to represent the deployment that took place.
- The **lead time** for that PR is calculated and added to the deployment as part of the blueprint.

Here’s how you can implement this:

1. **Define a Pull Request Blueprint**:
    - A sample PR blueprint can be found [here](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/git/github/examples/resource-mapping-examples).

> 2. **Use Lead Time Calculation**:

Reviewer: The lead time for PR should already be part of the guide you mentioned for Github. If not, let's include it in the default mapping and the guide.

    port:
      entity:
        mappings:
          identifier: ".head.repo.name + '-' + (.id|tostring)"
          title: "Deployment for PR {{ .head.repo.name }}"
          blueprint: '"deployment"'
          properties:
          environment: '"Production"' #Hard coded for now
          createdAt: ".merged_at"
          deploymentStatus: '"Successful"' #Hard coded for now
          leadTime: |
            (.created_at as $createdAt | .merged_at as $mergedAt | 
            ($createdAt | sub("\\..*Z$"; "Z") | strptime("%Y-%m-%dT%H:%M:%SZ") | mktime) as $createdTimestamp | 
            ($mergedAt | if . == null then null else sub("\\..*Z$"; "Z") | strptime("%Y-%m-%dT%H:%M:%SZ") | mktime end) as $mergedTimestamp | 
            if $mergedTimestamp == null then null else ($mergedTimestamp - $createdTimestamp) / 86400 end)
  
   ```
  </details>
  :::tip hardcoded values
    The value for deploymentStatus is hardcoded to `Successful` to treat all deployment as success
    and environment is hardcoded to production for main/master in this example.
    You can modify these values based on your requirements.
  :::

> 3. **Deployment Creation**:

Reviewer: Maybe a better approach is to update the Github mapping?

e.g. for a kind of pull-request, create a new deployment entity with search criteria based on the branch?

Author: @Matanga1-2 does this mean the automation be removed 🙂 ?

Reviewer: @kodjomiles yes 
The automation can be removed, and instead we can show an addition to Github mapping that generates the deployment entity

---
sidebar_position: 10
displayed_sidebar: null
---

import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"
import PortTooltip from "/src/components/tooltip/tooltip.jsx";

# DORA metrics

This guide is designed
to help you
understand and implement [DevOps Research and Assessment (DORA) metrics](https://cloud.google.com/devops/research/dora) within your organization in Port.
DORA Metrics are a set of key performance indicators
that measure the effectiveness and efficiency of your software development and delivery process. 
By tracking these metrics,
you can identify areas for improvement and ensure that your team is delivering high-quality software efficiently.
This guide will cover the four key metrics: Deployment Frequency, Lead Time, Change Failure Rate, and Mean Time to Recovery.

:::info Prerequisites
- Complete the [Port onboarding process](https://docs.getport.io/quickstart).

> - Access a GitHub repository (e.g., `port-dora-metrics`) with workflows or pipelines for tracking deployments. Follow the [GitHub integration guide](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/git/github) to sync deployments, pull requests, and workflows to Port.

Reviewer: I think we can remove it since it's not required for the basic/recommended approach.
You could write it below in the section that actually requires this

> - Access a GitHub repository (e.g., `port-dora-metrics`) with workflows or pipelines for tracking deployments. Follow the [GitHub integration guide](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/git/github) to sync deployments, pull requests, and workflows to Port.

Author: okay

> - Access a GitHub repository (e.g., `port-dora-metrics`) with workflows or pipelines for tracking deployments. Follow the [GitHub integration guide](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/git/github) to sync deployments, pull requests, and workflows to Port.

Author: can we do something like this 

- Complete the [Port onboarding process](https://docs.getport.io/quickstart).
- Access a GitHub repository (e.g., port-dora-metrics) to sync your pull requests or deployments. You can follow the [GitHub integration guide](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/git/github) to ensure your deployments are tracked in Port.
- Optional for advanced strategies: If you're using workflows or pipelines, ensure they are configured for deployment tracking by following the relevant setup guides, such as CI/CD or GitHub Actions.

> - Access a GitHub repository (e.g., `port-dora-metrics`) with workflows or pipelines for tracking deployments. Follow the [GitHub integration guide](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/git/github) to sync deployments, pull requests, and workflows to Port.

Author: make it an optional prerequisite ?

- When a PR is merged, a **deployment entity** is created in Port to represent the deployment that took place.
- The **lead time** for that PR is calculated and added to the deployment as part of the blueprint.

Here’s how you can implement this:

1. **Add Pull Request blueprint, sample can be found [here](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/git/github/examples/resource-mapping-examples/#map-repositories-and-pull-requests)**:

2. **Add the configuration below** to the [data sources page](https://app.getport.io/settings/data-sources) in your Port portal, and select your GitHub integration:

<details>
<summary><b>Deployment config (click to expand)</b></summary>

```yaml showLineNumbers
- kind: pull-request
  selector:
    query: .base.ref == 'main'  # Track PRs merged into the main branch
  port:
    entity:
      mappings:
        identifier: .head.repo.name + '-' + (.id|tostring)
        title: Deployment for PR {{ .head.repo.name }}
        blueprint: '"deployment"'
        properties:
          environment: '"Production"'  # Hard coded for now
          createdAt: .merged_at
          deploymentStatus: 'Success'  # Hard coded for now

>           leadTime: |

Author: it's a calculation property. taking that from the pr level would have to be mirrored in the deployment blueprint yes ?
Should we go for that instead ?

>           leadTime: |

Reviewer: Yeah a mirror property make sense, or alternatively map something like

leadTime: leadTime

>           leadTime: |

Author: we can't map leadTime to a leadTime from PR because that property doesn't exist.. it's a calculated property, making it a mirror would make sense.

>           leadTime: |

Author: and the lead time on the pr blueprint is in days not in minutes, should I still go ahead

>           leadTime: |

Author: "mirrorProperties": {
"leadTimeDays": {
"title": "Lead Time (Days)",
"mirror": ".relations.pullRequest.properties.lead_time_days",
"type": "number",
"description": "Mirrored lead time from Pull Request Blueprint in days."
}
},
"calculationProperties": {
"leadTimeMinutes": {
"title": "Lead Time (Minutes)",
"type": "number",
"calculation": ".properties.leadTimeDays * 1440",
"description": "Lead time in minutes, calculated from leadTimeDays."
}
}

and if that's the case we have to do a calculation to probably bring it to minutes

>           leadTime: |

Reviewer: @kodjomiles good questions
Lets finalize it later when we conclude the doc
For now just mirror a lead in days property.

>           leadTime: |

Author: okay

            (.created_at as $createdAt | .merged_at as $mergedAt | 
            ($createdAt | sub("\\..*Z$"; "Z") | strptime("%Y-%m-%dT%H:%M:%SZ") | mktime) as $createdTimestamp | 
            ($mergedAt | if . == null then null else sub("\\..*Z$"; "Z") | strptime("%Y-%m-%dT%H:%M:%SZ") | mktime end) as $mergedTimestamp | 
            if $mergedTimestamp == null then null else ($mergedTimestamp - $createdTimestamp) / 60 end)
```

</details>

:::tip Hardcoded values
The value for deploymentStatus is hardcoded to `Success` to treat all deployments as successful,
and the environment is hardcoded to Production for the **main** branch in this example.
You can modify these values based on your requirements.
:::

</TabItem>

<TabItem value="workflow-job" label="Workflow/Job">

Track deployments by monitoring workflow runs in your pipeline.
This setup captures all workflow runs from the main branch and maps them to deployment entities.
The deployment status is set dynamically based on whether the workflow run concluded successfully or failed.

Here’s how you can implement this:

> - **Add the configuration below** to the [data sources page](https://app.getport.io/settings/data-sources) in your Port portal, and select your GitHub integration:

Reviewer: @kodjomiles worth explaining that this can be similarly implement for other runs/workflows in other tools

        blueprint: '"deployment"'
        properties:
          environment: '"Production"'  # Set environment based on branch (main/master as Production)
          createdAt: .run_started_at
          deploymentStatus: .conclusion
          leadTime: |
            (.run_started_at as $startTime | .updated_at as $endTime |
            ($startTime | sub("\\..*Z$"; "Z") | strptime("%Y-%m-%dT%H:%M:%SZ") | mktime) as $startTimestamp |
            ($endTime | sub("\\..*Z$"; "Z") | strptime("%Y-%m-%dT%H:%M:%SZ") | mktime) as $endTimestamp |
            ($endTimestamp - $startTimestamp) / 60)
```
</details>

:::tip Using regex
You can use regex to track only workflows related to deployments
(e.g., workflows containing the word "deploy").
The environment is automatically determined based on the branch, where the main branch corresponds to Production.
:::

</TabItem>

<TabItem value="ci-cd-pipelines" label="CI/CD Pipelines">

CI/CD pipelines, such as those run by Jenkins, provide a robust way to track deployments. Jenkins, in particular, allows you to create and update entities in Port dynamically using Port's API as part of the pipeline execution.

**Example**:

> In Jenkins, each time a build is triggered or completed,

Reviewer: @kodjomiles do you think it's feasible to add examples for all the supported CI/CD tools?
Maybe a "details" section for each type?

Jenkins
Octopus
CircleCI
Azure PIpelines
Codefresh
Gitlab Pipelines

Also depends on your estimate

Author: This would make it very long

  - kind: release
    selector:
      query: "true"  
    port:
      entity:
        mappings:
          identifier: .release.name
          title: Deployment for Release {{ .release.name }}
          blueprint: '"deployment"'  
          properties:
            environment: '"Production"'  
            createdAt: .release.created_at
            deploymentStatus: '"Success"'  
          relations:
            tag: .release.tag_name
            service: .repo.name
  - kind: tag
    selector:
      query: "true"  
    port:
      entity:
        mappings:
          identifier: .tag.name
          title: Tag {{ .tag.name }}
          blueprint: '"tag"'  
          properties:
            commit_sha: .commit.sha
          relations:
            service: .repo.name

```

</details>

:::tip
This configuration maps the repository, release, and tag information to deployment entities in Port.
You can find more details about setting up GitHub integrations for repositories, releases,
and tags [here](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/git/github/examples/resource-mapping-examples/#map-repositories-repository-releases-and-tags).
:::

</TabItem>

</Tabs>

> 

Reviewer: @kodjomiles maybe worth adding one final example of "Custom API" explaining that you can always interact directly with Port API from your internal tools, and share ean example API call snippet to create a new deployment?

### Monorepo Tracking

Port doesn’t provide a specific solution for monorepos,
but **custom integrations** can be used to effectively track services or components within a monorepo.

- **Define blueprints** for each service or component inside the monorepo. Each microservice or feature can have unique properties such as name, deployment status, and dependencies.

- **Ingest data** via Port’s API by setting up logic to track changes, deployments, or PR merges for individual services within the monorepo. Use selectors to filter and map the relevant parts of the monorepo to your service blueprints.

Example configuration for a monorepo with multiple services:

```yaml
resources:
  - kind: service
    selector:

>       query: ".repository.name == 'monorepo' && (.service == 'service-A' || .service == 'service-B')"

Reviewer: @kodjomiles can you please explain this line and how it assist with monorepo?

Author: ### Purpose of this Query for Monorepos:

- **`.repository.name == 'monorepo'`**: This condition ensures that the mapping applies only to repositories named **'monorepo'**, which is typical in monorepos where multiple services live in a single repository.
  
- **`(.service == 'service-A' || .service == 'service-B')`**: This additional condition filters the specific services within the monorepo. For example, if the monorepo contains multiple services (e.g., **service-A** and **service-B**), this query isolates and maps those specific services only, ensuring that they are treated as separate entities in Port.

### How it Assists with Monorepos:
In a **monorepo** setup, you have multiple services housed within a single repository. Each service might have its own deployment pipeline, ownership, and lifecycle. This query allows Port to treat each service as a separate entity while still being sourced from the same monorepo. 

By mapping specific services using this query, we ensure that Port can differentiate between the services and track their specific metadata, deployments, and incidents separately. This is crucial for environments that require granular tracking and management of microservices or different features within a single repository.

    port:
      entity:
        mappings:
          identifier: .service
          title: .service
          blueprint: '"service"'
          properties:
            url: .html_url
            defaultBranch: .default_branch
```

:::tip Why Custom Integration?
If your monorepo setup includes multiple services or components,
**Port’s custom integration API** provides the flexibility
to tailor tracking and deployment updates to each component individually.
This allows granular visibility and management of each microservice in your monorepo.
:::

## Tracking Incidents

> Incidents are key to calculating important DevOps metrics, including **Change Failure Rate (CFR)** and **Mean Time to Recovery (MTTR)**. Effective incident tracking provides insights into how frequently deployments result in failures and how quickly these failures are resolved.

Reviewer: ```suggestion
Incidents are vital for DORA metrics, including **Change Failure Rate (CFR)** and **Mean Time to Recovery (MTTR)**. Effective incident tracking provides insights into how frequently deployments result in failures and how quickly these failures are resolved.
```

- [ServiceNow](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/servicenow)
- [Statuspage](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/statuspage)

For example, **OpsGenie** and **ServiceNow** allow you to map incidents to services similarly to **PagerDuty**.
The process involves syncing the incidents, creating relations between services,
and ensuring consistent data capture for CFR and MTTR.

### Relating Incidents to Services

Once the incident data is synced, you'll want to create a relation between incidents and services in your Port Builder.
This allows incidents caused by deployment failures to be linked directly to the impacted services,
making it easier to track recovery times and overall service health.

To relate incidents to services:

1. **Add a Relation Between Incident and Service**: Create a relation in Port Builder between your **Incident Blueprint** and the **Service Blueprint**. This links each incident to the service it affects, ensuring that failures caused by deployment issues are captured and tracked.

> 2. **Create a Mirror Property**: Set up mirror properties to reflect key incident data (e.g., **recoveryTime**, **resolvedAt**) on the related service. This makes it easier to track MTTR at the service level.

Reviewer: @kodjomiles maybe this can be improved and shorten a bit?
For example
"Add the below addition to your Incident blueprint to relate Incident and Service

Author: Straight to the point yeah okay

<details>
<summary><b>Relation and Mirror Property JSON (click to expand)</b></summary>

```json
{
  "relations": {
    "pagerdutyService": {
      "title": "Related Service",
      "target": "service",
      "many": false
    }
  },

>   "mirrorProperties": {

Reviewer: @kodjomiles why do we need mirror properties? are these mirrored from Incident into a Service?
Given a service has multiple incidents, what is the value of having it mirrored? 
Maybe it should be aggregated to the service level instead?

Author: if it's syncing across multiple incidents then aggregating is better

    "recoveryTime": {
      "title": "Recovery Time",
      "mirror": ".properties.recoveryTime"
    },
    "resolvedAt": {
      "title": "Resolution Time",
      "mirror": ".properties.resolvedAt"
    }
  }
}
```
</details>

> ### Metric Calculations

Reviewer: @kodjomiles is it possible to just give the bottom line?

For example, "We will now add the metrics to the service level"
<< service blueprint updates>>

"We will add the metrics to a new Organization level blueprint
<<organization blueprint>>
"We will manually create a new Organization entity and connect service/teams to it (maybe automation/SS action. should discuss"

That's it, your portal is not tracking DORA metrics. All we are left to do is visualize it in a Dashboard

Author: when can we discuss this

Reviewer: Discussed

> ### Metric Calculations

Reviewer: Also it should probably be ##? it's a new section, not under incidents

Author: thought we were putting it under incidents

Reviewer: @kodjomiles the overall structure can be

Deployment 
Incidents
Metrics
Dashboards

Agree?

Author: agree

To sync incidents from **PagerDuty**, follow the steps in the [PagerDuty guide](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/pagerduty). The guide provides detailed steps for setting up integrations to track incidents related to deployments.

For other incident management tools, follow these respective guides:
- [OpsGenie](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/opsgenie)
- [FireHydrant](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/firehydrant)
- [ServiceNow](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/servicenow)
- [Statuspage](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/statuspage)

### Relating Incidents to Services

To track incidents in relation to services, create a **relation** between the **Incident Blueprint** and the **Service Blueprint** in Port Builder. This allows incidents to be linked to services affected by deployment failures, enabling better monitoring of **MTTR** and overall service health.

> Instead of mirroring properties like `recoveryTime` and `resolvedAt` at the service level, **aggregate** them to calculate the average **MTTR** across multiple incidents.

Author: I have resolved this already, I think the comment was made on an old review

<details>
<summary><b>Relation and Aggregation Property JSON (click to expand)</b></summary>

```json
{
  "relations": {
    "pagerdutyService": {
      "title": "Related Service",
      "target": "service",
      "many": false
    }
  },
  "aggregationProperties": {
    "meanTimeToResolve": {
      "title": "Mean Time to Recovery",
      "target": "incident",
      "calculationSpec": {
        "calculationBy": "entities",
        "func": "avg",
        "field": "recoveryTime"
      },
      "query": {
        "combinator": "and",
        "rules": [
          {
            "property": "status",
            "operator": "=",
            "value": "resolved"
          }
        ]
      }
    }
  }
}
```
</details>

## Metrics

We will demonstrate how to calculate the four main DORA metrics using aggregated data:

1. **Deployment Frequency**: Number of successful deployments within a timeframe.
2. **Change Lead Time**: Time from PR opened to deployment.
3. **Change Failure Rate (CFR)**: Incidents divided by total deployments.
4. **Mean Time to Recovery (MTTR)**: Time from the incident's onset to resolution.

> To achieve this, we will aggregate data at the **Organization** level.

Author: this too

We'll create an **Organization** blueprint and add the necessary relationships for **Deployments**,
**PRs**, **Incidents**, and **Services**.

### Data Model Setup

> Define the **Organization** blueprint and add these relationships to connect it with Deployments,

Author: This isn't present aas well, I don't have anything like this anymore.
I think the review was done on old code

Pull Requests, Incidents, and Services:

<details> 
<summary><b>Organization Blueprint (click to expand)</b></summary>

```json
{
  "identifier": "organization",
  "title": "Organization",
  "schema": {
    "properties": {},
    "required": []
  },
  "relations": {
    "deployment": {
      "title": "Related Deployment",
      "target": "deployment",
      "many": true
    },
    "pull_request": {
      "title": "Related Pull Request",
      "target": "githubPullRequest",
      "many": true
    },
    "incident": {
      "title": "Related Incident",
      "target": "incident",
      "many": true
    },
    "service": {
      "title": "Related Service",
      "target": "service",
      "many": true
    }
  }
}
```
</details>

:::tip Why define these relationships?
These relationships enable the **Organization** blueprint
to pull data from the other relevant blueprints such as **Deployment**,
**Pull Request**, **Incident**, and **Service**.
Without these relationships,
the **Organization** entity wouldn’t be able to aggregate the required data to calculate the DORA metrics.
:::

> 

Author: I am confused here

      "title": "Average Deployments",
      "type": "number",
      "target": "deployment",
      "calculationSpec": {
        "func": "avg",
        "property": "deployment_frequency",
        "calculationBy": "property"
      }
    },
    "average_lead_time": {
      "title": "Average Lead Time (PR)",
      "type": "number",
      "target": "githubPullRequest",
      "calculationSpec": {
        "func": "avg",
        "property": "lead_time_days",
        "calculationBy": "property"
      }
    },
    "total_incidents": {
      "title": "Total Incidents",
      "type": "number",
      "target": "incident",
      "calculationSpec": {
        "func": "count",
        "calculationBy": "entities"
      }
    },
    "failure_rate": {
      "title": "Change Failure Rate",
      "type": "number",
      "calculation": "(count(.relations.incident) / count(.relations.deployment)) * 100"
    },
    "mean_time_to_recovery": {
      "title": "Mean Time to Recovery",
      "type": "number",
      "target": "incident",
      "calculationSpec": {
        "func": "avg",
        "property": "recoveryTime",
        "calculationBy": "property"
      }
    }
  }
}
```
</details>

> By defining the **Organization** blueprint, adding the necessary relationships, and using these calculation blocks, you can successfully calculate the four main DORA metrics: **Deployment Frequency**, **Change Lead Time**, **Change Failure Rate (CFR)**, and **Mean Time to Recovery (MTTR)**.

Author: We are no more dealing with organization, but I can still add that it takes 15 minutes to show up as part of the limitations which is already stated in the aggregation docs page.
